### PR TITLE
tls: add SNI support

### PR DESCRIPTION
**Pull request by alltilla**
Inspired by a syslog-ng maillist request:

> Hi,
> I am using TLS over TCP connection to forward my syslog events to a remote server.
> My remote server uses SNI (Server Name Identification) to route connections/events to one of the available backend servers.
> I observe that syslog-ng doesn't send SNI during TLS handshake.
> How can I enable it?
> My configuration is as follows:

```
source s_net { syslog(transport(udp) port(1514)); };
destination d_tcp {
        tcp(
                "XX.example.net"
                port(96)
                tls(
                        peer-verify(required-untrusted)
                        ca_dir("/etc/syslog-ng/ssl")
                        key-file("/etc/syslog-ng/ssl/globaltest/XX.example.net.key.pem")
                        cert-file("/etc/syslog-ng/ssl/globaltest/XX.example.net.cert.pem")
                  )
        );
};
log {
        source(s_net);
        destination(d_tcp);
};
```

> I want syslog-ng to send XX.example.net as SNI to my remote server
> Please advise
> Thanks
> Raghu

Example usage:
```
destination d_tls {
    tcp(
        "XX.example.net"
        port(96)
        tls(
            sni(yes)
            ca_dir("/etc/syslog-ng/ssl")
            key-file("/etc/syslog-ng/ssl/globaltest/XX.example.net.key.pem")
            cert-file("/etc/syslog-ng/ssl/globaltest/XX.example.net.cert.pem")
        )
    );
};
```
Wireshark:
client_hello

Signed-off-by: Attila Szakacs <attila.szakacs@balabit.com>

### CODE REVIEW

File: lib/tlscontext.c [hunk 1]
```diff
@@ -999,6 +1023,13 @@ tls_context_set_dhparam_file(TLSContext *self, const gchar *dhparam_file)
   self->dhparam_file = g_strdup(dhparam_file);
 }
 
+void
+tls_context_set_sni(TLSContext *self, const gchar *sni)
```
Comment (MrAnno) [inline lib/tlscontext.c]: This needs a declaration in the header.

File: lib/tlscontext.c [hunk 2]
```diff
@@ -313,6 +314,21 @@ tls_session_info_callback(const SSL *ssl, int where, int ret)
...
+static gboolean
+_set_sni_in_client_mode(TLSSession *self)
+{
+  if (!self->ctx->sni)
+    return TRUE;
+
+  if (self->ctx->mode != TM_CLIENT)
+    return TRUE;
+
+  if (SSL_set_tlsext_host_name(self->ssl, self->ctx->sni))
```
Comment (MrAnno) [inline lib/tlscontext.c]: Note that `SSL_set_tlsext_host_name()` is documented only for OpenSSL 1.1.1:
https://www.openssl.org/docs/man1.1.1/man3/SSL_get_servername_type.html

It exists in 1.0.2 and 1.1.0 without docs, so I guess it's OK, no compat layer is needed.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/tlscontext.c (hunk 1 above)
Comment (alltilla) [reply lib/tlscontext.c]: Yes, I forgot to commit :)

File: lib/tlscontext.c (hunk 2 above)
Comment (alltilla) [reply lib/tlscontext.c]: Yes, that is what I have seen. It is a fairly old code, not sure why the documentation was behind so much,

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
https://tools.ietf.org/html/rfc6066#page-6

File: lib/tlscontext.c [hunk 3]
```diff
@@ -326,6 +346,13 @@ tls_session_new(SSL *ssl, TLSContext *ctx)
 
   SSL_set_info_callback(ssl, tls_session_info_callback);
 
+  if (!_set_sni_in_client_mode(self))
+    {
+      tls_context_unref(self->ctx);
+      g_free(self);
+      return NULL;
```
Comment (MrAnno) [inline lib/tlscontext.c]: If this fails, a `SSL` object will be leaked every `time-reopen()` seconds.

```
324,918 (244,920 direct, 79,998 indirect) bytes in 39 blocks are definitely lost in loss record 97 of 97
   at 0x483877F: malloc (vg_replace_malloc.c:299)
   by 0x4C5F59A: CRYPTO_zalloc (in /usr/lib/libcrypto.so.1.1)
   by 0x4A92485: SSL_new (in /usr/lib/libssl.so.1.1)
   by 0x48A9551: tls_context_setup_session (tlscontext.c:771)
   by 0x5D71C8A: _construct_tls_transport (transport-mapper-inet.c:100)
   by 0x5D71DFC: transport_mapper_inet_construct_log_transport (transport-mapper-inet.c:145)
   by 0x5D6BFA2: transport_mapper_construct_log_transport (transport-mapper.h:83)
   by 0x5D6C896: afsocket_dd_construct_transport (afsocket-dest.c:213)
   by 0x5D6C9BC: afsocket_dd_connected (afsocket-dest.c:231)
   by 0x5D6CD8C: _afsocket_dd_connection_in_progress (afsocket-dest.c:290)
   by 0x499CCDF: ??? (in /usr/lib/libivykis.so.0.5.6)
   by 0x499E293: iv_main (in /usr/lib/libivykis.so.0.5.6)
```

Comment (alltilla) [reply lib/tlscontext.c]: Thanks, fixed! :)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**